### ISSUE TITLE
Recipe ideas

### ISSUE THREAD

**Issue by tbreloff** [tbreloff/Plots.jl#319]

**Comment by ChrisRackauckas** [comment]
What about column scatter plots? They are pretty widely used in Biology since it's like a boxplot in that shows an error bar, but is also plotting the actual data values. At the bottom of this is a good example (made in R):

![13227_2016_40_fig12_html](https://cloud.githubusercontent.com/assets/1814174/15992153/e77428de-307a-11e6-8ed3-177f693bb5b6.jpg)

**Comment by diegozea** [comment]
In that omage, also the upper bars with the significance level are useful and generally hard to make, but should be easy to make a recipe for that significance annotations with Plots.

**Comment by ChrisRackauckas** [comment]
If adding significance annotations is possible that was be fantastic, but I thought it would be hard (it would be easy to tell it that a certain plot is *** significant, but then for it to automatically place the annotations? That seems difficult).

**Comment by tbreloff** [comment]
No. Annotations should be fairly straightforward (depending on what it is
of course). You'd just create a new series with 'primary = false'.

**Comment by tbreloff** [comment]
For anyone watching the repo, I just wanted to give a quick demo of features I just added: `iter_segments` and the `curves` series recipe.

![tmp](https://cloud.githubusercontent.com/assets/933338/16088975/491adae6-32f7-11e6-8c6a-8aa263340d6c.png)

If you can't figure out by the screenshot, I added a nice convenience iterator `iter_segments`, which takes 1 or more vectors (or anything that can be "cycled", which includes scalar values), and gives an iterator that, at each iteration, returns an integer range which is the next consecutive set of indices where the inputs are all non-NaN.  I'll be replacing things like `gr_polyline` with this sort of iteration logic, though it has other uses:

The `:curves` series recipe will split the x/y/z/fillrange/line_z coordinates into segments, and rebuild each segment using the given coordinates as "control points" in a bezier curve.  So you can create nice curves simply by specifying the control points separated by NaN's.  The optional `npoints` keyword arg specifies how many line segments to use to draw each curve segment.

**Comment by mkborregaard** [comment]
Rather than choropleths, or as a way of doing that, it could be nice with a recipe to plot shapefiles, i.e. the objects returned by Shapefile.jl. Allowing colour would make an easy way to do choropleths, in addition to the other uses of shapefiles.

**Comment by tbreloff** [comment]
This should be pretty straightforward. Can we add a recipe to shapefile.jl?

**Comment by mkborregaard** [comment]
Don't know, but I have the impression they welcome PRs. I made some super rudimentary headway:
```
using Plots, Shapefile

shp = open("test.shp") do fd
    read(fd, Shapefile.Handle)
end
pol = shp.shapes[1] # there may be more than one polygon, but in this case only one

xs = ys = Float64[]
for i in 1:pol.points
    push!(xs, i.x)
    push!(ys, i.y)
end

#xs and ys become identical  - this seems to be a bug, so I had to do a workaround
n = length(xs)/2
ys = xs[2*(1:n)]
xs = xs[2*(1:n)-1]

plot(xs, ys, aspect_ratio = :equal, xlim = (-70, -60)) 
```
This gives a line plot, but what we want for this type of shapefile is a polygon - cannot find in the docs how to do filled polygons.

**Comment by tbreloff** [comment]
Add 'st = :shape'! Coordinates can be NaN-separated to have multiple shapes
in the same series. I wonder if this should be added to Shapefile or
PlotRecipes?

**Comment by mkborregaard** [comment]
Oh, nice trick with the NaN to get multiple shapes, but I cannot seem to get different colours on the shape [pic](https://www.dropbox.com/s/notob86fj50m0q5/shapefile.png?dl=0)
```
plot([xs; NaN; xs + 3], [ys; NaN; ys], aspect_ratio = :equal, xlim = (-70, -60), st = :shape, colour = [:green :red], legend = false)
```
By the way, the shape file is [here](https://www.dropbox.com/sh/ksdncn7tntdqscm/AACZxHKh4gt-wzKkEs8DgM_Ha?dl=0)

I think it would make sense to put it in Shapefile, since it depends 100% on the functionality of that package.

**Comment by tbreloff** [comment]
Right now the easiest way to set different colors per shape is to make each
one a different series. Eventually this will be easier.

**Comment by mkborregaard** [comment]
Nice, `plot([xs (xs + 3)], [ys ys], aspect_ratio = :equal, st = :shape, colour = [:green :red], legend = false)` works. I'll make a recipe and do a PR on Shapefile, see what the owners think.

**Comment by mkborregaard** [comment]
Ah, no, of course making them different series does not work as the polygons will not have the same number of points. I tried using groups to get different colours, but that takes prohibitively long (I stopped my machine after a few minutes). I also tried a for loop with plot! but I could not get it to input to screen. So I will wait a bit with implementing this till I figured out how to do the shapes in different colours.

**Comment by tbreloff** [comment]
This is actually easier than you're making it.  There are a few improvements to be made to colors and shapes, after which this should work well:

![tmp](https://cloud.githubusercontent.com/assets/933338/16653994/b9913cee-4421-11e6-9c62-83b155c080f2.png)

**Comment by mkborregaard** [comment]
> This is actually easier than you're making it

Story of my life, actually :-)

Anyway, that is amazing. I didn't know about the Shape class, but even if I did I would not have been able to do that so elegantly. I learn a lot from reading it. And it works! 

![skaermbillede 2016-07-07 kl 15 19 39](https://cloud.githubusercontent.com/assets/8429802/16654485/4fa8c38a-4456-11e6-8efc-c3bdbfe4f47d.png)

**Comment by tbreloff** [comment]
:+1: What shape file is that?  All the test files seem to be very simple polygons.

**Comment by mkborregaard** [comment]
That is one I am working on at the moment, of 1500 islands in the IndoPacific. It is not mine so I cannot put it online but I will PM it to you if you want it?

**Comment by tbreloff** [comment]
Oh... no need.  Do you know of any good sources for other maps/shapes with a permissive license.

**Comment by tbreloff** [comment]
I think you should use this instead:

```julia
Base.convert(::Type{Plots.Shape}, s::Shapefile.ESRIShape) = Shape([pt.x for pt=s.points], [pt.y for pt=s.points])
@recipe f{T<:Shapefile.ESRIShape}(::Type{T}, shape::T) = Shape(shape)
@recipe f{T<:AVec{Shapefile.ESRIShape}}(::Type{T}, shapes::T) = Shape[s for s=shapes]
@recipe f{T<:AMat{Shapefile.ESRIShape}}(::Type{T}, shapes::T) = Shape[s for s=shapes]'
```

So that you can do:

![tmp](https://cloud.githubusercontent.com/assets/933338/16654913/f33d8610-4425-11e6-8cdc-5b51a408e825.png)

**Comment by mkborregaard** [comment]
>  Do you know of any good sources for other maps/shapes with a permissive license.
Yes, the Natural Earth site is excellent, and all data is in the public domain (no credits necessary). Take e.g. this:
[](http://www.naturalearthdata.com/downloads/110m-physical-vectors/110m-coastline/)

**Comment by tbreloff** [comment]
Cool!  It would be pretty awesome to have a package which downloads/unzips this data and wraps loading through Shapefile.jl.

**Comment by mkborregaard** [comment]
👍

**Comment by mkborregaard** [comment]
Hey that 4-line code is fantastic, thanks. I will try and generalize it to Line and Point shapefiles as well. There is a hitch in the plotting of the shapefile I suggested, Plots gets confused in a few places about what is outside or inside. This may be a bug in Shapefiles.jl, though.
<img width="626" alt="screen shot 2016-07-07 at 21 21 06" src="https://cloud.githubusercontent.com/assets/8429802/16666600/c15b984a-4488-11e6-9cb3-f0cc710b75e0.png">

**Comment by tbreloff** [comment]
If there are "jumps" in the same shape it will do this.  I suspect the shapes need to be broken up... I can't think of a good way to automate that without domain knowledge of what the shape represents.

**Comment by mkborregaard** [comment]
I am not going to pretend to know what is going on. The problem before arose because I missed the transpose operator. Remembering it, I get this:
<img width="617" alt="screen shot 2016-07-07 at 21 28 52" src="https://cloud.githubusercontent.com/assets/8429802/16666868/d8002a24-4489-11e6-9c7b-93689c585a0f.png">
Seriously, WTF? :-)

**Comment by mkborregaard** [comment]
Sorry - I think it is a line shapefile. My bad.

**Comment by tbreloff** [comment]
> Seriously, WTF? :-)

:laughing: 

So they aren't actually polygons... they're partial borders.  That explains the weirdness.

**Comment by mkborregaard** [comment]
No there truly is something strange going on. I opened a new session and only typed this:
<img width="740" alt="screen shot 2016-07-07 at 21 49 21" src="https://cloud.githubusercontent.com/assets/8429802/16667455/dada70e4-448c-11e6-9b31-89af9b1ce060.png">

And now I know for sure that it is polygons that plot correctly (with fill colour) in R.

**Comment by tbreloff** [comment]
what happens if you add `st = :shape` to that command?

**Comment by mkborregaard** [comment]
That, and specifying `color` instead of `fill`, helps a lot!
<img width="616" alt="screen shot 2016-07-07 at 22 00 44" src="https://cloud.githubusercontent.com/assets/8429802/16667758/588e4b9a-448e-11e6-8262-8ceff703fe37.png">

**Comment by tbreloff** [comment]
getting closer!  the border isn't perfect, but I don't know if there's a Plots-solution for that.

**Comment by mkborregaard** [comment]
No, I am not sure there is. It may be a slight difference in polygon ending convention among shapefile polygons and other shapes. I was just looking into whether I could define the ticks, st, aspect_ratio and legend in a plot recipe to make plotting easier - but it looks like that would require defining a specialized function, right?

**Comment by tbreloff** [comment]
you can!  here's a "plot recipe" to add to your list:

```julia
@recipe function f(::Type{Val{:shapeplot}}, plt::Plot)
    legend := false
    ticks := nothing
    grid := false
    aspect_ratio := 1
    seriestype := :shape
    ()
end
@shorthands shapeplot
```

and then (with some bonus features that only I possess):

![tmp](https://cloud.githubusercontent.com/assets/933338/16670036/5ed055de-4466-11e6-8aeb-7f3dee897f46.png)

If we figure out how to extract closed polygons properly, then we're done :+1:

**Comment by mkborregaard** [comment]
Ha ha, here is the one I just did (but it does have a specialized function name):
`
@recipe function f(s::Shapeplot)
    seriestype := :shape
    aspect_ratio := :equal
    grid --> false
    legend --> false
    ticks --> :none
    s.args
end
`

**Comment by tbreloff** [comment]
that'll work too :+1:

**Comment by mkborregaard** [comment]
My plot seems to have fewer problems, that is weird.
<img width="618" alt="screen shot 2016-07-07 at 23 27 35" src="https://cloud.githubusercontent.com/assets/8429802/16670423/75e08cec-449a-11e6-9585-10188c0ab2f8.png">

**Comment by tbreloff** [comment]
what commit are you on?  i might have introduced a bug

**Comment by mkborregaard** [comment]
0.7.0 here on julia 0.4.5, I'm on my home machine. Do you want me to `Pkg.update()`?

**Comment by mkborregaard** [comment]
Try and do `shapeplot(shp.shapes')` instead of `shapeplot(shp.shapes)`

**Comment by tbreloff** [comment]
Just for fun:

```julia
plot(xlim=(-200,200), ylim=(-100,100))
@gif for s in shp.shapes
    shapeplot!(s, c=:grey)
end
```

![tmp](https://cloud.githubusercontent.com/assets/933338/16673875/204e05a8-4480-11e6-9b4f-24f296a57abe.gif)

I thought there might have been a bug in the `iter_segments` logic, but I think it's ok.  Not sure why some of those polygons are weird.

**Comment by tbreloff** [comment]
Hmm shapes 94 and 95 should really be combined:

![tmp](https://cloud.githubusercontent.com/assets/933338/16674043/247c89f4-4482-11e6-9219-d20e505ae609.png)

**Comment by tbreloff** [comment]
Maybe easiest to get from: https://github.com/nvkelso/natural-earth-vector

**Comment by tbreloff** [comment]
For reference, this is probably the nicest way to grab the files:

```
fn = "https://github.com/nvkelso/natural-earth-vector/raw/master/packages/Natural_Earth_quick_start/10m_cultural/ne_10m_admin_1_states_provinces.shp"
run(`wget $fn -P /tmp/`)
```

Sadly they don't seem to be quite robust enough to use out of the box.  I'll revisit another time.

**Comment by mkborregaard** [comment]
I think the issue here is you used the old one (the Lines shapefile with the coastline). I did my plots on the newer shapefile (the Polygons type). Here is how it looks:
![skaermbillede 2016-07-08 kl 11 15 30](https://cloud.githubusercontent.com/assets/8429802/16682964/85e5ad12-44fd-11e6-9eb9-61f8041371fe.png)
Shapefile recipes need to be specialized on the type of shapefile.
I don't think there is anything wrong with the natural earth layers, if there are issues anywhere it is probably in the Shapefile.jl package.
Would be nice to do a NaturalEarth package, yes. Does the wget command also work in Windows, though?

**Comment by mkborregaard** [comment]
There are still some kinks to be ironed out but I think we could make a PR. Do you want to do, as you wrote the code essentially, or should I? And do you want it on PlotRecipes or to Shapefile?

**Comment by tbreloff** [comment]
How about this... I'll add the shapeplot recipe to PlotRecipes, and you can
submit the Shapefile-specific type recipes to Shapefile.jl. If they don't
want it then I'll add it to PlotRecipes with an "is_installed guard".

**Comment by mkborregaard** [comment]
Sure, let us say that.

**Comment by mkborregaard** [comment]
Will that require Shapefile.jl to depend on Plots.jl because of the Plots.Shape type?

**Comment by tbreloff** [comment]
Ah. Good point. I'll rewrite them to avoid using Shape.

**Comment by lobingera** [comment]
@mkborregaard https://github.com/tbreloff/Plots.jl/issues/319#issuecomment-231315455
I'd assume to not to have wget on windows (any revision) out of the box. 
But https://github.com/JuliaWeb/Requests.jl should provide something (system independend)

**Comment by mkborregaard** [comment]
👍

**Comment by mkborregaard** [comment]
> I'll rewrite them to avoid using Shape.

Or I could put something like: `isdefined(:Plots) && isa(Plots, Module) && include("Plotsrecipe.jl")` in the Shapefile module, and then putting the `import Plots: Shape` statement in the Plotsrecipe.jl file. I think that should keep it from adding the dependency?

**Comment by tbreloff** [comment]
No... I'll rewrite them. The shapes are immediately converted to x/y
coordinate vectors anyways... It's a wasteful step.

**Comment by mkborregaard** [comment]
Ah-ha, and suddenly my unnecessarily complicated way of doing things may attain renewed prominence 😺
I was going to add an extra step regardless, because some Shapefile polygons are more than one shape. That is encoded in the parts fieldname. This means that every polygon should be converted to an array of shapes, that then gets splatted into Plots. An example of this is already in the Shapefile package, for Compose shapes: https://github.com/JuliaGeo/Shapefile.jl/blob/master/src/compose.jl

Let me know what you want me to do here. I am happy that this is getting implemented, especially when the different colours also gets into the release.

**Comment by mkborregaard** [comment]
Could be easily converted to
```julia
function convert(::Type{AVec{Shape}},
        shape::Shapefile.Polygon)
    points = {}
    polygons={}
    currentpart=2
    for (i,p) in enumerate(shape.points)
        push!(points, p)
        if i==length(shape.points) || (currentpart≤length(shape.parts) && i==shape.parts[currentpart])
            push!(polygons, Shape([(p.x,p.y) for p in points]))
            currentpart += 1
            points = {}
        end
    end
    polygons
end
```

**Comment by tbreloff** [comment]
@mkborregaard Can you confirm what the `shape.parts` is?  Does that break up a `Shapefile.Polygon` into several sub-polygons?  Is `parts` the same length as `points`, or is it a list of "indices of breaks"?

**Comment by mkborregaard** [comment]
`Polygon.parts` is an `Int32[]` . It has the same length as the number of shapes in each Polygon. The first element is always 0. The number is the index in `Polygon.points` just before the index starting each new point. So, eg if there are 5 parts with a total of 420 points, parts may be [0, 9, 34, 73, 315].
The code I have pasted above should work, though.

**Comment by tbreloff** [comment]
Ok I'm just going to do this slightly differently so I can't copy/paste that code block.  I'll post the recipes soon.

**Comment by mkborregaard** [comment]
👍

**Comment by tbreloff** [comment]
Almost done... which file are you using?  I'd like to download the same one as here: https://github.com/tbreloff/Plots.jl/issues/319#issuecomment-231315455

**Comment by mkborregaard** [comment]
http://www.naturalearthdata.com/downloads/110m-physical-vectors/110m-land/

**Comment by tbreloff** [comment]
Ok great.  For reference, here are the recipes:

To be added to Shapefile.jl:

```julia
using RecipesBase
function shapefile_coords(poly::Shapefile.ESRIShape)
    start_indices = poly.parts+1
    end_indices = vcat(poly.parts[2:end], length(poly.points))
    x, y = zeros(0), zeros(0)
    for (si,ei) in zip(start_indices, end_indices)
        push!(x, NaN)
        push!(y, NaN)
        for pt in poly.points[si:ei]
            push!(x, pt.x)
            push!(y, pt.y)
        end
    end
    x, y
end

function shapefile_coords{T<:Shapefile.ESRIShape}(polys::AbstractVector{T})
    x, y = zeros(0), zeros(0)
    for poly in polys
        xpart, ypart = shapefile_coords(poly)
        append!(x, xpart)
        append!(y, ypart)
    end
    x, y
end

function shapefile_coords{T<:Shapefile.ESRIShape}(polys::AbstractMatrix{T})
    x, y = [], []
    for c in 1:size(polys,2)
        xy = shapefile_coords(vec(polys[:,c]))
        push!(x, xy[1])
        push!(y, xy[2])
    end
end

@recipe f(poly::Shapefile.ESRIShape) = (seriestype --> :shape; shapefile_coords(poly))
@recipe f{T<:Shapefile.ESRIShape}(polys::AbstractVector{T}) = (seriestype --> :shape; shapefile_coords(polys))
@recipe f{T<:Shapefile.ESRIShape}(polys::AbstractMatrix{T}) = (seriestype --> :shape; shapefile_coords(polys))
```

To be added to PlotRecipes:

```julia
@recipe function f(::Type{Val{:shapeplot}}, plt::Plot)
    legend --> false
    ticks --> nothing
    grid --> false
    aspect_ratio --> 1
    seriestype := :shape
    ()
end
@shorthands shapeplot
```

A sample script:

```julia
using PlotRecipes, Shapefile
dir = "https://github.com/nvkelso/natural-earth-vector/raw/master/110m_physical/"
fn = "ne_110m_land.shp"
run(`wget $dir/$fn -P /tmp/`)
shp = open("/tmp/$fn") do fd
    read(fd, Shapefile.Handle)
end
shapeplot(shp.shapes, c=:grey)
```

And some silliness:

![tmp](https://cloud.githubusercontent.com/assets/933338/16688521/6c45aae4-44ec-11e6-949d-d63cac43e000.gif)

**Comment by mkborregaard** [comment]
Awesome! Do you want me to put those in the PR? It feels a bit strange, since you wrote this completely.

Just a few questions - it looks like you use NaN separation for the shapes from the same polygon. Does that mean the will retain the same attributes? E.g. will it be possible to pass a vector of colours or annotations (e.g. names) to the shapeplot function with one element per original Polygon and get the right colors out? (necessary for choropleths).

Also, does the sample script work? It shouldn't be possible to just get the .shp file, there are at least 4 different files in a shapefile.

**Comment by tbreloff** [comment]
> Awesome! Do you want me to put those in the PR? It feels a bit strange, since you wrote this completely.

I don't mind... I'd rather not spend the time putting the PR together.

> E.g. will it be possible to pass a vector of colours or annotations (e.g. names) to the shapeplot function with one element per original Polygon and get the right colors out?

This is something that is working on the `tb_colors` branch of Plots... you'd call it like: `shapeplot(shapes, c = my_color_list)`.  The one caveat is that the list would treat each "part" as its own polygon, so you may need to handle this specially inside the shapefile code if you want to map a color to each of the Polygon parts.

> Also, does the sample script work? It shouldn't be possible to just get the .shp file, there are at least 4 different files in a shapefile.

When downloading from github, you can get just the `.shp` file without the `.zip`.

**Comment by mkborregaard** [comment]
I also think it would be good with a
`@recipe f{T<:Shapefile.Handle}(::Type{T}, handle::T) = handle.shapes'`
which could dispatch on handle.type - this code only works for type == 5 (there are 28 of them, but I think only 1, 3, and 5 are necessary).

**Comment by tbreloff** [comment]
Sure... feel free to add/expand!  I only wanted to provide the basics.

**Comment by mkborregaard** [comment]
> When downloading from github, you can get just the .shp file without the .zip.

I mean can you run the code with just the .shp file?
As @lobingera there is the julia alternative to wget
```julia
using Requests
tmp = get("https://github.com/nvkelso/natural-earth-vector/raw/master/110m_physical/ne_110m_land.shp")
save(tmp, "tmp.shp")
shp = open("tmp.shp") do fd
        read(fd, Shapefile.Handle)
    end
```

> Sure... feel free to add/expand! I only wanted to provide the basics.

Awesome, I'll do the PR and expand to the other types. I would say that is essential that multi-part polygons are treated as an entity in the plotting, I will look into that once I'm 100% clear what your code is doing.

**Comment by mkborregaard** [comment]
PR is here: https://github.com/JuliaGeo/Shapefile.jl/pull/10

**Comment by tbreloff** [comment]
`shapeplot` added to PlotRecipes: https://github.com/JuliaPlots/PlotRecipes.jl/commit/88a7a0f47ae26a79d3b8246e09474cdccecbba29

**Comment by ChrisRackauckas** [comment]
Should basically all of these go to PlotRecipes.jl and StatRecipes.jl?

**Comment by tbreloff** [comment]
@ChrisRackauckas yes

**Comment by tbreloff** [comment]
posting this to use in the readme:

![tmp](https://cloud.githubusercontent.com/assets/933338/16770876/83dea362-481c-11e6-9943-bb77148be5b8.png)

**Comment by AndyGreenwell** [comment]
Just now learning Plots myself, what would be involved in expanding the recipe used for `ohlc` to also overlay various average trendlines? [Example from wikipedia](https://en.wikipedia.org/wiki/Open-high-low-close_chart#/media/File:Bollinger_bands_example,_2_stddevs.png)

**Comment by tbreloff** [comment]
Should be pretty easy to add yourself. May I recommend OnlineStats.Variance
for the tracking?

**Comment by djsegal** [comment]
Bump for waterfall 🌊

**Comment by kellertuer** [comment]
Bump for a 3D quiver
```
using Plots
plotly()
x = [1.0, 2.0, 3.0]
y = [0.0, 0.0, 0.0]
z = y
u = [1.0, 1.0, 1.0]
v = [0.0, 1.0, 0.0]
w = [1.0, 0.0, 1.0]
plot(x,y,z, quiver=(u,v,w), seriestype=:quiver)
```
just ignores `z` and `w` (also on `gr()`).
